diff --git a/qemu_mode/qemu-8.2.0/system/physmem.c b/qemu_mode/qemu-8.2.0/system/physmem.c
index a63853a7b..ff94a7e83 100644
--- a/qemu_mode/qemu-8.2.0/system/physmem.c
+++ b/qemu_mode/qemu-8.2.0/system/physmem.c
@@ -1883,8 +1883,10 @@ static void ram_block_add(RAMBlock *new_block, Error **errp)
          * KVM is not used and it may be forked (eg for fuzzing purposes).
          */
         if (!qtest_enabled()) {
+            /* Keep translated memory blocks across forks for AFL!
             qemu_madvise(new_block->host, new_block->max_length,
                          QEMU_MADV_DONTFORK);
+            */
         }
         ram_block_notify_add(new_block->host, new_block->used_length,
                              new_block->max_length);
