diff --git a/qemu_mode/qemu-8.2.0/qemu-options.hx b/qemu_mode/qemu-8.2.0/qemu-options.hx
index 42fd09e4d..8ba116ac4 100644
--- a/qemu_mode/qemu-8.2.0/qemu-options.hx
+++ b/qemu_mode/qemu-8.2.0/qemu-options.hx
@@ -4106,6 +4106,13 @@ SRST
     from ./my\_blob.bin.
 ERST
 
+DEF("aflFile", HAS_ARG, QEMU_OPTION_aflFile, \
+    "-aflFile fname  AFL input sourced from fname\n", QEMU_ARCH_ALL)
+DEF("aflPanicAddr", HAS_ARG, QEMU_OPTION_aflPanicAddr, \
+    "-aflPanicAddr hexaddr  Address of OS panic function\n", QEMU_ARCH_ALL)
+DEF("aflDmesgAddr", HAS_ARG, QEMU_OPTION_aflDmesgAddr, \
+    "-aflDmesgAddr hexaddr  Address of OS logging function\n", QEMU_ARCH_ALL)
+
 DEF("serial", HAS_ARG, QEMU_OPTION_serial, \
     "-serial dev     redirect the serial port to char device 'dev'\n",
     QEMU_ARCH_ALL)
