diff --git a/qemu_mode/qemu-8.2.0/tcg/tcg.c b/qemu_mode/qemu-8.2.0/tcg/tcg.c
index 896a36cae..caab2e0f3 100644
--- a/qemu_mode/qemu-8.2.0/tcg/tcg.c
+++ b/qemu_mode/qemu-8.2.0/tcg/tcg.c
@@ -764,6 +764,14 @@ static void alloc_tcg_plugin_context(TCGContext *s)
 #endif
 }
 
+static void free_tcg_plugin_context(TCGContext *s)
+{
+#ifdef CONFIG_PLUGIN
+    g_ptr_array_unref(s->plugin_tb->insns);
+    g_free(s->plugin_tb);
+#endif
+}
+
 /*
  * All TCG threads except the parent (i.e. the one that called tcg_context_init
  * and registered the target's TCG globals) must register with this function
@@ -816,6 +824,21 @@ void tcg_register_thread(void)
 }
 #endif /* !CONFIG_USER_ONLY */
 
+void tcg_unregister_thread(void)
+{
+    TCGContext *s = tcg_ctx;
+    unsigned int n;
+
+    /* Unclaim an entry in tcg_ctxs */
+    n = qatomic_fetch_dec(&tcg_cur_ctxs);
+    g_assert(n > 0);
+    qatomic_store_release(&tcg_ctxs[n - 1], 0);
+
+    free_tcg_plugin_context(s);
+
+    g_free(s);
+}
+
 /* pool based memory allocation */
 void *tcg_malloc_internal(TCGContext *s, int size)
 {
